### ISSUE TITLE
Worker is unable to mantain connection with Controller located in another network

### ISSUE THREAD

**Issue by j0rzsh** [hashicorp/boundary#758]
**Describe the bug**

We have deployed an infrastructure based on: [boundary-reference-architecture](https://github.com/hashicorp/boundary-reference-architecture) but with several modifications including GCP infrastructure:

In AWS:

- Two public subnets and one private subnet
- One controller and one worker in a public subnet
- One target in the private subnet
- The LoadBalancer has one new target group pointing to 9201 port to controller

In GCP:

- One public subnet and one private subnet in GCP
- One worker in public subnet in GCP
- One target in the private subnet

We have zero problems when dealing with the AWS Infra, but the worker in GCP is unable to mantain connection to the controller in AWS. Configuration in GCP Instances is replicated from the AWS EC2.

In the following images, we can see that the GCP worker reach the AWS controller in the first place through the Load Balancer, but once the communication has been established, it is lost because the controller tells the worker to reach it using its private address (10.0.0.11 in this case). As both instances are in different network, this can't be accomplished (Journalctl is executed in the GCP worker instance):

![journalctl](https://user-images.githubusercontent.com/38765045/97581950-7d9c2880-19f5-11eb-94e0-83bc104bf840.png)

We have make it work executing the following command in the GCP worker instance:

```bash
iptables -t nat -A OUTPUT -d 10.0.0.11 -j DNAT --to-destination 54.239.32.132

## 10.0.0.11 is the private address of the controller, the IP that controller sends to worker
## 54.239.32.132 was the public IP address of the load balancer
```

Doing this, boundary works as expected, allowing us to ssh to the GCP target instance using boundary connect ssh.

We suppose that this is caused by the controller sending its private IP address. We think that a possible solution might be to separate private IP address to expose boundary service and external IP address to send to workers in different configuration parameters.

**To Reproduce**
Run boundary with configuration stated in [boundary-reference-architecture](https://github.com/hashicorp/boundary-reference-architecture) and add a worker in another network.

**Expected behavior**
Boundary Worker should be able to mantain the connection with the Controller.

**Comment by malnick** [comment]
Hi @j0rzsh - thanks for trying out Boundary. This is the first use case using a hybrid cloud scenario I've seen, really cool stuff. 

Regarding worker -> controller communication. The communication needs to be in both directions:
1. The worker needs to be able to connect to the controller to register itself.
2. The controller needs to be able to connect to the worker to handle session establishment for a given client.

I have not tried this, but I'm wondering if you'd be able to use the load balancer for the controller as the value for the worker `controllers[]` configuration?

**Comment by j0rzsh** [comment]
Hi @malnick. Thanks for the response!

Yes, we have tried both the public IP of the controller and the load balancer public endpoint in ```controllers[]``` configuration. Both tries gives us the same result: connection established correctly at first but when keeping alive this connection the worker receives the private IP of the controller and stops reaching it.

When the worker connects to the controller at the first step it uses the ```controllers[]``` configuration that we wrote in the config file, but then the controller connects back to the worker to handle session  and sends its own local IP (or the entire controllers list IPs) .

So we haven’t foud a way to tell the boundary controller not to use its local IP when it has to be accessed throught a public endpoint. If you see the current state of the server table in the database, the controller IP that it is stored is its local IP. In contrast, the workers stores theirs public IPs.

|private_id|type|name|description|address|create_time|update_time|
|----------|----|----|-----------|-------|-----------|-----------|
|demo-worker-gcp|worker|demo-worker-gcp|A default worker created for demonstration|35.241.209.144:9202|2020-10-30 09:18:42|2020-10-30 09:18:42|
|demo-worker-aws|worker|demo-worker-aws|A default worker created for demonstration|34.247.183.215:9202|2020-10-30 09:18:39|2020-10-30 09:55:04|
|demo-controller-0|controller|demo-controller-0|A controller for a demo!|10.0.0.8:9201|2020-10-30 09:17:10|2020-10-30 09:55:11|

If we can help further let us know!

**Comment by jefferai** [comment]
@j0rzsh I just merged #761 which gives you the ability to specify the cluster address that should be advertised to workers after they connect. That should solve your issue when it comes out in 0.1.2 (or you can build from main now!)

**Comment by j0rzsh** [comment]
Hi @malnick, @jefferai!
I'm just writing to say thanks. We built it from main and now it is working perfectly!
We will continue testing boundary and keeping track of it's new features.

Thanks again for your great work!
